### PR TITLE
Utilize pixman for efficient pixel manipulation

### LINKED ISSUE #6: Utilize pixman for portable and efficient pixel manipulation
[Pixman](https://gitlab.freedesktop.org/pixman/pixman) is a library that provides low-level pixel manipulation features such as image compositing and trapezoid rasterization. It is released under an MIT-style license, which perfectly matches the licensing terms of `Mado`. In the file `src/primitive.c`, various pixel manipulations are implemented in portable C, but most are inefficient due to the lack of hardware features in modern microprocessors, such as SIMD. By integrating Pixman, `Mado` can manipulate low-level pixels more efficiently.

### PR DESCRIPTION
**Pull request by Bennctu**
According to issue #6, I replace original image compositing and trapezoid rasterization with Pixman function. Briefly, include:
 - pixman_image_composite: image compositing
 - pixman_fill_rect: use pixman_image_fill_rectangles to fill the image with retangles
 - pixman_image_fill_boxes: span is filled with the boxs

Close #6

We compare the results of the original method with those rewritten using the Pixman library:
(Left image: original method, and right image: Pixman method)
- Hello world:

- Calculator(Original):

- Viewer(Original):

- Multi-demo(Original):

### CODE REVIEW

File: configs/Kconfig [hunk 1]
```diff
@@ -39,6 +39,10 @@ config CURSOR
     bool "Manipulate cursor"
     default n
 
+config PIXMAN
+    bool "Pixman to pixel manipulation"
+    default y
```
Comment (jserv) [inline configs/Kconfig:44]: Since it is experimental, don't set by default.

**Comment by jserv** [comment]
Consider the performance test program below (`apps/perf.c`)
```c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>
#include <time.h>

#include "twin.h"

#define TEST_PIX_WIDTH 1200
#define TEST_PIX_HEIGHT 800

static twin_pixmap_t *src32, *dst32;
static int twidth, theight, titers;

static void test_argb32_source_argb32(void)
{
    twin_operand_t srco = {.source_kind = TWIN_PIXMAP, .u.pixmap = src32};
    twin_composite(dst32, 0, 0, &srco, 0, 0, NULL, 0, 0, TWIN_SOURCE, twidth,
                   theight);
}

static void test_argb32_over_argb32(void)
{
    twin_operand_t srco = {.source_kind = TWIN_PIXMAP, .u.pixmap = src32};
    twin_composite(dst32, 0, 0, &srco, 0, 0, NULL, 0, 0, TWIN_OVER, twidth,
                   theight);
}

static void do_test(const char *name, void (*test)(void))
{
    struct timeval start, end;
    unsigned long long sus, eus;
    char spc[128];
    char *s;
    int i;

    printf("%s", name);

    gettimeofday(&start, NULL);
    for (i = 0; i < titers; i++)
        test();
    gettimeofday(&end, NULL);
    sus = (unsigned long long) start.tv_sec * 1000000ull + start.tv_usec;
    eus = (unsigned long long) end.tv_sec * 1000000ull + end.tv_usec;

    s = spc;
    for (i = strlen(name); i < 40; i++)
        *(s++) = ' ';
    *s = 0;
    printf("%s %f sec\n", spc, ((float) (eus - sus)) / 1000000.0);
}

#define DO_TEST(name) do_test(#name, test_##name)

static void do_tests(int width, int height, int iters)
{
    twidth = width;
    theight = height;
    titers = iters;

    DO_TEST(argb32_source_argb32);
    DO_TEST(argb32_over_argb32);
}

static void do_all_tests(const char *title)
{
    printf("[ %s: 10x10x1000000 ]\n", title);
    do_tests(10, 10, 1000000);

    printf("[ %s: 100x100x20000 ]\n", title);
    do_tests(100, 100, 20000);

    printf("[ %s: 200x200x10000 ]\n", title);
    do_tests(200, 200, 10000);

    printf("[ %s: 1200x800x200 ]\n", title);
    do_tests(1200, 800, 200);

    printf("\n");
}

int main(void)
{
    /* Create some test pixmaps */
    src32 = twin_pixmap_from_file("assets/tux.png", TWIN_ARGB32);
    assert(src32);
    dst32 = twin_pixmap_create(TWIN_ARGB32, TEST_PIX_WIDTH, TEST_PIX_HEIGHT);
    assert(dst32);

    /* fill pixmaps */
    twin_fill(dst32, 0x80112233, TWIN_SOURCE, 0, 0, TEST_PIX_WIDTH,
              TEST_PIX_HEIGHT);

    /* pre-touch data */
    test_argb32_source_argb32();

    do_all_tests("Pixmap");

    return 0;
}
```

Enable it by turning off the existing demo programs. (You might integrate the above into `apps/main.c` though.)
```diff
--- a/Makefile
+++ b/Makefile
@@ -109,7 +109,7 @@ endif
 ifeq ($(CONFIG_DEMO_APPLICATIONS), y)
 target-y += demo-$(BACKEND)
 demo-$(BACKEND)_depends-y += $(target.a-y)
-demo-$(BACKEND)_files-y = apps/main.c
+demo-$(BACKEND)_files-y = apps/perf.c
 demo-$(BACKEND)_includes-y := include
 demo-$(BACKEND)_ldflags-y := \
        $(target.a-y) \
```

Reference test results with the builtin compositing:
```
[ Pixmap: 10x10x1000000 ]
argb32_source_argb32                     0.130208 sec
argb32_over_argb32                       0.273309 sec
[ Pixmap: 100x100x20000 ]
argb32_source_argb32                     0.025439 sec
argb32_over_argb32                       0.431355 sec
[ Pixmap: 200x200x10000 ]
argb32_source_argb32                     0.087813 sec
argb32_over_argb32                       0.839746 sec
[ Pixmap: 1200x800x200 ]
argb32_source_argb32                     0.022176 sec
argb32_over_argb32                       0.395362 sec
```

**Comment by jserv** [comment]
Promote Pixman option in Kconfig:
```diff
--- a/configs/Kconfig
+++ b/configs/Kconfig
@@ -16,6 +16,18 @@ config BACKEND_SDL
 
 endchoice
 
+choice
+    prompt "Renderer Selection"
+    default RENDERER_BUILTIN
+
+config RENDERER_BUILTIN
+    bool "Built-in pixel manipulation"
+
+config RENDERER_PIXMAN
+    bool "Pixman based rendering"
+
+endchoice
+
 menu "Features"
 
 config LOGGING
@@ -39,10 +51,6 @@ config CURSOR
     bool "Manipulate cursor"
     default n
 
-config PIXMAN
-    bool "Pixman to pixel manipulation"
-    default y
-
 endmenu
 
 menu "Image Loaders"
```

Once the `RENDERER_PIXMAN` option is enabled, the built-in pixel manipulation code should not be compiled, and you should ensure that all rendering is done through the Pixman path.

File: apps/main.c [hunk 2]
```diff
@@ -47,6 +47,34 @@ static twin_pixmap_t *load_background(twin_screen_t *screen, const char *path)
         twin_pixmap_destroy(raw_background);
         return twin_make_pattern();
     }
+#if defined(CONFIG_PIXMAN)
```
Comment (jserv) [inline apps/main.c:50]: Don't do that. The rendering path to Pixman should be transparent to all applications using Mado, meaning there should be no changes required on the application side when Mado is configured to switch from the built-in pixel manipulation to the faster Pixman path.

File: src/widget.c [hunk 3]
```diff
@@ -54,10 +54,18 @@ void _twin_widget_paint_shape(twin_widget_t *widget,
...
-    if (shape == TwinShapeRectangle)
+    if (shape == TwinShapeRectangle) {
+#if defined(CONFIG_PIXMAN)
+        pixman_color_t color;
+        twin_argb32_to_pixman_color(widget->background, &color);
+        pixman_fill_rect(pixmap, &color, pixmap->clip.left, pixmap->clip.top,
+                         right, bottom);
+        twin_pixmap_damage(pixmap, left, top, right, bottom);
+#else
         twin_fill(pixmap, widget->background, TWIN_SOURCE, left, top, right,
```
Comment (ndsl7109256) [inline src/widget.c:65]: I think you could rewrite `twin_fill` as following to hide the detailed implementation
```c
void twin_fill(...)
{
#if defined(CONFIG_PIXMAN)
...pixman library implementaion
#else
...original implementation
#endif
}
```

**Comment by jserv** [comment]
Pixman is a low-level library for pixel manipulation, including image compositing and trapezoid rasterization. It serves as the backend for Cairo's in-memory rendering surface and is a binary dependency of Cairo. Mado can be seen as a simplified version of Cairo, and the proposed change introduces an efficient rendering path as an optional feature for faster vector graphics. The original pixel manipulation code should be retained for resource-constrained environments, with the potential for integration into the Linux kernel as an in-kernel vector graphics module.

The proposed change adds a configurable option to switch between the built-in (or classical) renderer and the Pixman-based renderer. Applications using Mado should maintain source-level compatibility. The action items for this pull request include:
1. Implement `pixman_image_composite`, `pixman_fill_rect`, `pixman_image_fill_rectangles`, etc., within Mado's internal implementation, ensuring no changes to the public API and maintaining source-level compatibility.
2. Develop a simple performance benchmark program to validate and measure the efficiency of pixel manipulations, enabling comparisons between the built-in and Pixman-based renderers. This task can be done in another pull request.
3. Ensure the conditional build system works as expected, generating the necessary renderer-specific code without unnecessary code bloat when configured.

File: Makefile [hunk 4]
```diff
@@ -62,6 +61,16 @@ libtwin.a_includes-y := \
 libtwin.a_files-$(CONFIG_LOGGING) += src/log.c
 libtwin.a_files-$(CONFIG_CURSOR) += src/cursor.c
 
+# Rendering
+ifeq ($(CONFIG_RENDERER_BUILTIN), y) ## Builtin
```
Comment (jserv) [inline Makefile:65]: The `ifeq` is not necessary. You can simply write as following:
```
libtwin.a_files-$(CONFIG_RENDERER_BUILTIN) += src/draw.c
```

Both `CONFIG_RENDERER_BUILTIN` and `CONFIG_RENDERER_PIXMAN` are mutually exclusive, meaning that it is not possible to build the two files at the same time. Thus, `ifeq` is redundant.

File: Makefile [hunk 5]
```diff
@@ -62,6 +61,16 @@ libtwin.a_includes-y := \
 libtwin.a_files-$(CONFIG_LOGGING) += src/log.c
 libtwin.a_files-$(CONFIG_CURSOR) += src/cursor.c
 
+# Rendering
```
Comment (jserv) [inline Makefile:64]: Refine the comment to "Rendering path."

File: README.md [hunk 6]
```diff
@@ -63,9 +63,9 @@ benefiting the entire application stack.
 `Mado` is built with a minimalist design in mind. However, its verification
 relies on certain third-party packages for full functionality and access to all
 its features. To ensure proper operation, the development environment should
-have the [SDL2 library](https://www.libsdl.org/), [libjpeg](https://www.ijg.org/), and [libpng](https://github.com/pnggroup/libpng) installed.
+have the [SDL2 library](https://www.libsdl.org/), [libjpeg](https://www.ijg.org/), [libpng](https://github.com/pnggroup/libpng), and [libpixman-1](https://pixman.org/) installed.
```
Comment (jserv) [inline README.md:66]: Honor the name. Use `pixman`.

File: src/pixman.c [hunk 7]
```diff
@@ -0,0 +1,212 @@
+/*
+ * Twin - A Tiny Window System
+ * Copyright (c) 2004 Keith Packard <keithp@keithp.com>
```
Comment (jserv) [inline src/pixman.c:3]: Change the copyright notice to
> Copyright (c) 2024 National Cheng Kung University, Taiwan

File: README.md [hunk 8]
```diff
@@ -63,9 +63,9 @@ benefiting the entire application stack.
...
 * macOS: `brew install sdl2 jpeg libpng`
```
Comment (jserv) [inline README.md:67]: For macOS, the package name listed in Homebrew is `pixman`.

**Comment by jserv** [comment]
Build failure with Clang:
```
src/poly.c:258:17: error: call to undeclared function '_span_fill'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  258 |                 _span_fill(pixmap, y, x0, a->x);
      |                 ^
1 error generated.
```

**Comment by jserv** [comment]
I am experiencing rendering precision issues when switching from the built-in renderer to the Pixman-based one.

- [ ] The built-in renderer
built-in renderer

- [ ] The Pixman-based renderer
pixman-based renderer

As shown in the diagrams above, the splines rendered using the Pixman-based path are not as accurate as those rendered with the built-in renderer.

**Comment by jserv** [comment]
Evaluate the feasibility to replace the primitive operations, such as `_twin_rgb16_source_argb32`, `_twin_argb32_over_argb32`, and `_twin_argb32_source_argb32`, with Pixman counterpart. Eventually, the file `src/primitive.c` can be omitted if Pixman-based render path is selected.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `main` branch.

File: Makefile [hunk 9]
```diff
@@ -62,6 +61,18 @@ libtwin.a_includes-y := \
...
 
+# Rendering path
+## Builtin
+libtwin.a_files-$(CONFIG_RENDERER_BUILTIN) += src/draw.c
+
+## Pixman
+libtwin.a_files-$(CONFIG_RENDERER_PIXMAN) += src/pixman.c
+libtwin.a_cflags-$(CONFIG_RENDERER_PIXMAN) += $(shell pkg-config --cflags pixman-1)
+ifeq ($(CONFIG_RENDERER_PIXMAN), y)
+	TARGET_LIBS += $(shell pkg-config --libs pixman-1)
```
Comment (jserv) [inline Makefile:72]: Don't indent here.

**Comment by Bennctu** [comment]
https://github.com/sysprog21/mado/pull/55#issuecomment-2394969749
It results from _span_fill in pixman.c. I don't add new _span_fill method which uses pixman function.

File: src/pixmap.c [hunk 10]
```diff
@@ -13,6 +13,10 @@ twin_pixmap_t *twin_pixmap_create(twin_format_t format,
                                   twin_coord_t height)
 {
     twin_coord_t stride = twin_bytes_per_pixel(format) * width;
+    /* Padding the stride to a multiple of 4 bytes */
```
Comment (jserv) [inline src/pixmap.c:16]: Why do you think it is necessary? Explain and consider to submit another pull request for generic pixmap manipulation.

**Comment by jserv** [comment]
> I don't add new _span_fill method which uses pixman function.

Show preliminary benchmark results for comparison purposes.

File: src/pixmap.c (hunk 10 above)
Comment (Bennctu) [reply src/pixmap.c:16]: If we comment it, it will happen _pixman_log_error as follow:
```
*** BUG ***
*** BUG ***
In create_bits_image_internal: The expression bits == NULL || (rowstride_bytes % sizeof (uint32_t)) == 0 was false
Set a breakpoint on '_pixman_log_error' to debug

[1]    62492 segmentation fault (core dumped)  ./demo-sdl
```

pixman_image_t *msk fails to be allocated in [/src/pixman.c#L102](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR102). Finally, pixman_image_unref will access a nullptr and happen segmentation fault in [/src/pixman.c#L107](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR107)

**Comment by Bennctu** [comment]
https://github.com/sysprog21/mado/pull/55#issuecomment-2395457975

- _span_fill with pixman function in [/src/pixman.c](https://github.com/sysprog21/mado/commit/9a9b5de9224263ea730548c02e88b8c1558e77d3#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR194)

- _span_fill with the builtin method in poly.c

(Notice that we test in the pixman configuration)

File: src/pixmap.c (hunk 10 above)
Comment (jserv) [reply src/pixmap.c:16]: Do not use screenshots for plain text content, as this is inaccessible to visually impaired users.

Comment (jserv) [reply src/pixmap.c:16]: > If we comment it, it will happen _pixman_log_error as follow:
> ...
> `pixman_image_t *msk` fails to be allocated. Finally, `pixman_image_unref` will access a nullptr, and segmentation fault occurs.

Next, you should check the pointer alignment requirements for Pixman and consider providing the corresponding macros for alignment. See [mimalloc](https://github.com/microsoft/mimalloc/blob/dev/include/mimalloc/internal.h#L308-L324) for reference.

File: Makefile [hunk 11]
```diff
@@ -62,6 +61,18 @@ libtwin.a_includes-y := \
 libtwin.a_files-$(CONFIG_LOGGING) += src/log.c
 libtwin.a_files-$(CONFIG_CURSOR) += src/cursor.c
 
+# Rendering path
+## Builtin
```
Comment (jserv) [inline Makefile:65]: The comment was redundant.

File: src/pixman.c [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+        twin_to_pixman_op(operator), dst, &color, 1,
+        &(pixman_rectangle16_t){left, top, right - left, bottom - top});
+
+    twin_pixmap_damage(_dst, left, top, right, bottom);
+
+    pixman_image_unref(dst);
+}
+
+/* Same function in draw.c */
+static twin_argb32_t _twin_apply_alpha(twin_argb32_t v)
```
Comment (jserv) [inline src/pixman.c:151]: See if the function can be accelerated by Pixman.

**Comment by Bennctu** [comment]
https://github.com/sysprog21/mado/pull/55#issuecomment-2394989348
> Evaluate the feasibility to replace the primitive operations, such as , , and , with Pixman counterpart. Eventually, the file can be omitted if Pixman-based render path is selected.`_twin_rgb16_source_argb32``_twin_argb32_over_argb32``_twin_argb32_source_argb32``src/primitive.c`

According to [pixman.h](https://gitlab.freedesktop.org/pixman/pixman/-/blob/master/pixman/pixman.h?ref_type=heads), the in or over operator don't exist in public API. Because `_twin_rgb16_source_argb32`, `_twin_argb32_over_argb32`, and `_twin_argb32_source_argb32` are lower-level functions, we can only call pixman_image_composite (which can replace twin_composite in draw.c) to use in or over operation.

File: src/pixman.c (hunk 12 above)
Comment (Bennctu) [reply src/pixman.c:151]: Observing the difference between using (left image) and not using (right image) _twin_apply_alpha, we know that _twin_apply_alpha performs the conversion from ABGR to ARGB. 
imageimage
According to the implementation in _twin_apply_alpha,
```c
    return alpha << 24 | twin_int_mult(twin_get_8(v, 0), alpha, t1) << 16 |
           twin_int_mult(twin_get_8(v, 8), alpha, t2) << 8 |
           twin_int_mult(twin_get_8(v, 16), alpha, t3) << 0;
```
it also multiplies RBG channels with alpha value.

However, the public API in [pixman.h](https://gitlab.freedesktop.org/pixman/pixman/-/blob/master/pixman/pixman.h?ref_type=heads) doesn't provide the functions about ABGR to ARGB conversion and alpha premultiplication for user.

Comment (jserv) [reply src/pixman.c:151]: > we know that `_twin_apply_alpha` performs the conversion from ABGR to ARGB.

Check the inner function `_convertBGRtoARGB` in file `src/image-png.c` which was a workaround for macOS.

File: src/pixmap.c [hunk 13]
```diff
@@ -4,15 +4,33 @@
  * All rights reserved.
  */
 
+#include <assert.h>
 #include <stdlib.h>
 
 #include "twin_private.h"
 
+#define ALIGNMENT sizeof(uint32_t) /* the size of pixman image bits*/
```
Comment (jserv) [inline src/pixmap.c:12]: `sizeof(uint32_t)` is always 32 by definition.

File: src/pixman.c (hunk 12 above)
Comment (Bennctu) [reply src/pixman.c:151]: I ran mado with gdb. The function _convertBGRtoARGB isn't executed because my operating system is not macOS.

[/src/pixman.c#L172](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR172) is executed because my computer is a little-endian machine. The PNG data is originally in RGBA format as the following image, but it becomes ABGR format on a little-endian machine. In [/src/pixman.c#L172](https://github.com/sysprog21/mado/commit/7ccf36edd9a40450add80dc154c345d2d66d2dfd#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR172), mado has already performed the ABGR-to-ARGB conversion for the requirement of the pixmap's pixel data. 
 
 
image

File: src/pixmap.c [hunk 14]
```diff
@@ -4,15 +4,33 @@
...
 
+#define ALIGNMENT 4 /* the size (unit: bytes) of pixman image pixels*/
+#define IS_ALIGNED(p, alignment)            \
+    ({                                      \
+        assert(alignment != 0);             \
+        (((uintptr_t) p % alignment) == 0); \
+    })
+#define ALIGN_UP(sz, alignment)                                              \
+    ({                                                                       \
+        assert(alignment != 0);                                              \
```
Comment (jserv) [inline src/pixmap.c:20]: It is unlikely to have zero alignment. Drop assertion.

File: src/pixmap.c [hunk 15]
```diff
@@ -4,15 +4,28 @@
  * All rights reserved.
  */
 
+#include <assert.h>
```
Comment (jserv) [inline src/pixmap.c:7]: Drop the inclusion of `<assert.h>` since it is no longer used.

File: src/pixmap.c [hunk 16]
```diff
@@ -8,11 +8,22 @@
 
 #include "twin_private.h"
 
+#define IS_ALIGNED(p, alignment) ({ (((uintptr_t) p % alignment) == 0); })
```
Comment (jserv) [inline src/pixmap.c:11]: Don't use GNU extension `({` and `})`.

**Comment by jserv** [comment]
> According to [pixman.h](https://gitlab.freedesktop.org/pixman/pixman/-/blob/master/pixman/pixman.h?ref_type=heads), the in or over operator don't exist in public API. Because `_twin_rgb16_source_argb32`, `_twin_argb32_over_argb32`, and `_twin_argb32_source_argb32` are lower-level functions, we can only call pixman_image_composite (which can replace twin_composite in draw.c) to use in or over operation.

Can you provide preliminary benchmark results for evaluation purpose?

**Comment by Bennctu** [comment]
> > According to [pixman.h](https://gitlab.freedesktop.org/pixman/pixman/-/blob/master/pixman/pixman.h?ref_type=heads), the in or over operator don't exist in public API. Because `_twin_rgb16_source_argb32`, `_twin_argb32_over_argb32`, and `_twin_argb32_source_argb32` are lower-level functions, we can only call pixman_image_composite (which can replace twin_composite in draw.c) to use in or over operation.
> 
> Can you provide preliminary benchmark results for evaluation purpose?

The following results come from [this test code](https://github.com/sysprog21/mado/pull/55#issuecomment-2387338230):
```bash
# built-in method
[ Pixmap: 10x10x1000000 ]
argb32_source_argb32                     0.076773 sec
argb32_over_argb32                       0.130076 sec
[ Pixmap: 100x100x20000 ]
argb32_source_argb32                     0.065616 sec
argb32_over_argb32                       0.176866 sec
[ Pixmap: 200x200x10000 ]
argb32_source_argb32                     0.147732 sec
argb32_over_argb32                       0.352873 sec
[ Pixmap: 1200x800x200 ]
argb32_source_argb32                     0.058365 sec
argb32_over_argb32                       0.458145 sec
```
```bash
# pixman
[ Pixmap: 10x10x1000000 ]
argb32_source_argb32                     0.162651 sec
argb32_over_argb32                       0.149407 sec
[ Pixmap: 100x100x20000 ]
argb32_source_argb32                     0.020530 sec
argb32_over_argb32                       0.053927 sec
[ Pixmap: 200x200x10000 ]
argb32_source_argb32                     0.051592 sec
argb32_over_argb32                       0.132560 sec
[ Pixmap: 1200x800x200 ]
argb32_source_argb32                     0.041629 sec
argb32_over_argb32                       0.058690 sec
```

**Comment by jserv** [comment]
> The following evaluation results come from [this test code](https://github.com/sysprog21/mado/pull/55#issuecomment-2387338230):
> 
> ```shell
> # built-in method
> [ Pixmap: 10x10x1000000 ]
> argb32_source_argb32                     0.076773 sec
> argb32_over_argb32                       0.130076 sec
> # pixman
> [ Pixmap: 10x10x1000000 ]
> argb32_source_argb32                     0.162651 sec
> argb32_over_argb32                       0.149407 sec
> ```

Can you explain the test item `10x10x1000000`?

**Comment by Bennctu** [comment]
> > The following evaluation results come from [this test code](https://github.com/sysprog21/mado/pull/55#issuecomment-2387338230):
> > ```shell
> > # built-in method
> > [ Pixmap: 10x10x1000000 ]
> > argb32_source_argb32                     0.076773 sec
> > argb32_over_argb32                       0.130076 sec
> > # pixman
> > [ Pixmap: 10x10x1000000 ]
> > argb32_source_argb32                     0.162651 sec
> > argb32_over_argb32                       0.149407 sec
> > ```
> 
> Can you explain the test item `10x10x1000000`?

The following results show that it results from the allocation of pixman_image_t for src and dst in ['twin_composite' of pixman.c](https://github.com/sysprog21/mado/commit/54b51b57a651acab63200edc4cbd9c384c7da52c#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR64):
1. **In small compositing size**, the variation of the time consumption for pixman method is less than built-in method:
```
#Pixman
[ Pixmap: 7x7x1000000 ]
argb32_source_argb32                     0.169647 sec
argb32_over_argb32                       0.144932 sec
[ Pixmap: 9x9x1000000 ]
argb32_source_argb32                     0.151711 sec
argb32_over_argb32                       0.150564 sec
[ Pixmap: 11x11x1000000 ]
argb32_source_argb32                     0.169914 sec
argb32_over_argb32                       0.160988 sec
[ Pixmap: 13x13x1000000 ]
argb32_source_argb32                     0.172802 sec
argb32_over_argb32                       0.163671 sec
[ Pixmap: 15x15x1000000 ]
argb32_source_argb32                     0.194813 sec
argb32_over_argb32                       0.191646 sec
[ Pixmap: 17x17x1000000 ]
argb32_source_argb32                     0.183014 sec
argb32_over_argb32                       0.186015 sec
//////////////////////////////////////////////////////
#builtin
[ Pixmap: 7x7x1000000 ]
argb32_source_argb32                     0.054997 sec
argb32_over_argb32                       0.089470 sec
[ Pixmap: 9x9x1000000 ]
argb32_source_argb32                     0.070971 sec
argb32_over_argb32                       0.121374 sec
[ Pixmap: 11x11x1000000 ]
argb32_source_argb32                     0.088222 sec
argb32_over_argb32                       0.162219 sec
[ Pixmap: 13x13x1000000 ]
argb32_source_argb32                     0.112656 sec
argb32_over_argb32                       0.224784 sec
[ Pixmap: 15x15x1000000 ]
argb32_source_argb32                     0.130276 sec
argb32_over_argb32                       0.281983 sec
[ Pixmap: 17x17x1000000 ]
argb32_source_argb32                     0.161015 sec
argb32_over_argb32                       0.345266 sec
```

2. We find that the allocation of src and dst consumes **0.06 sec** (with iters = 1000000) approximately if we comment on ['twin_composite' of pixman.c](https://github.com/sysprog21/mado/commit/54b51b57a651acab63200edc4cbd9c384c7da52c#diff-5f7bbe04dc9ae37abc7b25930f4f97c723459d3c7791e04ec99804d4baf98e6fR64)
```
[ Pixmap: 7x7x1000000 ]
argb32_source_argb32                     0.073393 sec
argb32_over_argb32                       0.067797 sec
[ Pixmap: 9x9x1000000 ]
argb32_source_argb32                     0.074549 sec
argb32_over_argb32                       0.062425 sec
[ Pixmap: 11x11x1000000 ]
argb32_source_argb32                     0.062554 sec
argb32_over_argb32                       0.063226 sec
```

3. When compositing with smaller size, the allocation time of src and dst is larger proportion in total time consumption. Only doing pixman_image_composite in twin_composite result is close to built-in method:
```
[ Pixmap: 10x10x1000000 ]
argb32_source_argb32                     0.076398 sec
argb32_over_argb32                       0.078906 sec
[ Pixmap: 100x100x20000 ]
argb32_source_argb32                     0.020046 sec
argb32_over_argb32                       0.054756 sec
[ Pixmap: 200x200x10000 ]
argb32_source_argb32                     0.050169 sec
argb32_over_argb32                       0.133911 sec
[ Pixmap: 1200x800x200 ]
argb32_source_argb32                     0.046361 sec
argb32_over_argb32                       0.059155 sec
```

Generally, the results with pixman method are better than the built-in method.

**Comment by jserv** [comment]
> Generally, the results with pixman method are better than the built-in method.

We should integrate the performance suite before merging the proposed Pixman-based renderer to analyze its performance behavior and identify any regressions. Please submit a separate pull request for that purpose.

**Comment by jserv** [comment]
I plan to merge this pull request for further performance evaluation. Meanwhile, I would create new issue(s) to address pixman specific work. Some of them would be as following:
* Evaluate the feasibility to replace the primitive operations, such as `_twin_rgb16_source_argb32`, `_twin_argb32_over_argb32`, and `_twin_argb32_source_argb32`, with Pixman counterpart.
*  the file `src/primitive.c` can be omitted if Pixman-based render path is selected.
* Benchmarking! Locate the performance hotspot.

**Review by jserv — CHANGES_REQUESTED** [review]
Refine the git commit message to reflect the implementation aspects. In addition, place `Close #6` at the end of git commit message.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `main` branch, resolve conflicts, and highlight the switch between the built-in renderer and the Pixman-based implementation. For RISC-V platforms without vector or SIMD extensions, the built-in renderer offers simplicity and works well. On modern CPU architectures with rich ISA extensions, Pixman provides faster pixel manipulation, making the switch to the Pixman-based implementation advantageous.

**Comment by Bennctu** [comment]
> Rebase the latest `main` branch, resolve conflicts, and highlight the switch between the built-in renderer and the Pixman-based implementation. For RISC-V platforms without vector or SIMD extensions, the built-in renderer offers simplicity and works well. On modern CPU architectures with rich ISA extensions, Pixman provides faster pixel manipulation, making the switch to the Pixman-based implementation advantageous.

Could I refer to part of your sentence as a reference and include it in the commit message?

**Comment by jserv** [comment]
> Could I refer to part of your sentence as a reference and include it in the commit message?

Sure! It is collaboration.

**Review by jserv — CHANGES_REQUESTED** [review]
Don't say "According to issue #6." This pull request was linked to that already.

Don't say "I replace original image compositing with Pixman function." Always write down commit message in third-person point of view.

**Review by jserv — CHANGES_REQUESTED** [review]
Show (partial) benchmark results in git commit messages.

**Comment by jserv** [comment]
Thank @Bennctu for contributing!
